### DISCUSSION TITLE
Middle points in the pickup and delivery problem: provide some points in which the vehicles can interchange load

### DISCUSSION THREAD

**Discussion by marcojerome** [google/or-tools#2409] [Routing (and legacy CP) questions]
Hi, 

It would be nice to have the possibility to notify some additional points to the solver in which the vehicles can exchange load. In this way the pickups and deliveries could be done collaboratively among vehicles. 

**What language and solver does this apply to?**
Python

**Describe the problem you are trying to solve.**
VRP with time windows, capacity, pickup and delivery

**Describe the solution you'd like**
I did not find a way to code it in or-tools, but it might be possible. 

Thank you

**Comment by Mizux** [comment]
So It means for a Pickup&Delivery A,B you have two possibilities:

- Having the route `A ->...-> B` (i.e. P&D by the same vehicle) or
  - `time_dimension.CumulVar(A) < time_dimension.CumulVar(B)`
  - `load(A) = +1`, `load(B) = -1`
  -  `VehicleVar(A) == VehicleVar(B)` if C0 and C1 are not visited (which can be written `1 - ActiveVar(C0)*ActiveVar(C1)`)
- Having two routes `A -> ... -> C0`  and `C1 -> ... -> B` (i.e. P&D by two vehicles) with
  - `load(A) = +1`, `load(B) = -1`
  - `time_dimension.CumulVar(A) < time_dimension.CumulVar(B)`
  - `load(C0) = -1`, `load(C1) = +1`
  - if C0 and C1 are visited (aka `ActiveVar(C0) * ActiveVar(C1) == 1`) then `time_dimension.CumulVar(C0) < time_dimension.CumulVar(C1)`
  - if C0 and C1 are visited then
    -   `VehicleVar(A) == VehicleVar(C0)` and `VehicleVar(C1) == VehicleVar(B)`
 
So visiting `[C0, C1]` is optional and it is both or nothing which can be expressed as:
```python
c0 = mange.NodeToIndex(C0)
c1 = mange.NodeToIndex(C1)
routing.AddDisjunction([c0, c1], 0, 2)
```
